Firestore/core/src/firebase/firestore/remote/stream.h

```diff
@@ -0,0 +1,247 @@
+   * be in a non-started state, requiring the caller to start the stream again
+   * before further use.
+   *
+   * Only streams that are in state 'Open' can be marked idle, as all other
+   * states imply pending network operations.
+   */
+  void MarkIdle();
+
+  /** Marks the stream as active again. */
+  void CancelIdleCheck();
+
+ protected:
+  // `Stream` expects all its methods to be called on the Firestore queue.
+  void EnsureOnQueue() const;
+  void Write(grpc::ByteBuffer&& message);
+  void ResetBackoff();
+  std::string GetDebugDescription() const;
+
+ private:
+  /**
+   * `Stream` can be in one of 5 states (each described in detail below)
+   * shown in the following state transition diagram:
+   *
+   *          Start() called             auth & connection succeeded
+   * INITIAL ----------------> STARTING -----------------------------> OPEN
+   *                             ^  |                                   |
+   *                             |  |                    error occurred |
+   *                             |  \-----------------------------v-----/
+   *                             |                                |
+   *                    backoff  |                                |
+   *                    elapsed  |              Start() called    |
+   *                             \--- BACKOFF <---------------- ERROR
+   *
+   * [any state] --------------------------> INITIAL
+   *               Stop() called or
+   *               idle timer expired
+   */
+  enum class State {
+    /**
+     * The stream is not yet running and there's no error condition.
+     * Calling `Start` will start the stream immediately without backoff.
+     * While in this state, `IsStarted` will return false.
+     */
+    Initial,
+
+    /**
+     * The stream is starting, either waiting for an auth token or for the
+     * stream to successfully open. While in this state, `IsStarted` will
+     * return true but `IsOpen` will return false.
+     */
+    Starting,
+
+    /**
+     * The stream is up and running. Requests and responses can flow
+     * freely. Both `IsStarted` and `IsOpen` will return true.
+     */
+    Open,
+
+    /**
+     * The stream encountered an error. The next start attempt will back off.
+     * While in this state, `IsStarted` will return false.
+     */
+    Error,
+
+    /**
+     * An in-between state after an error where the stream is waiting before
+     * re-starting. After waiting is complete, the stream will try to open.
+     * While in this state, `IsStarted` will return true but `IsOpen` will
+     * return false.
+     */
+    Backoff
+  };
+
+  // The interface for the derived classes.
+
+  virtual std::unique_ptr<GrpcStream> CreateGrpcStream(
+      Datastore* datastore, absl::string_view token) = 0;
+  // PORTING NOTE: equivalent to `tearDown`.
+  virtual void FinishGrpcStream(GrpcStream* stream) = 0;
+  virtual void DoOnStreamStart() = 0;
```

Comment:
Shouldn't these be protected rather than private?

Also it seems as if the relationship between the Stream and callbacks in implementations could be made portable to the other platforms.

On Android Stream is parameterized on a separate `CallbackT` type. There it's generic on the request and response type but we don't need to do so here.

In Typescript, there's a `ListenerType` that is similar.

We could align the C++ implementation on similarly. Perhaps `StreamListener`?

Reply:
Making these protected would allow the derived classes to invoke these functions, but actually only the base class needs to invoke them. They are intended to be customization points.

On one hand, I find that `Stream` already has too many responsibilities to add another one to it. Moreover, notifying a delegate is a responsibility that cannot be fully implemented by `Stream`, so this logic becomes spread between the base class and the derived classes, which I find harder to grasp. On the other hand, there's consistency with other platforms. Would you mind if I do this change in the follow-up PR that deals with `WatchStream`?

Reply:
Update: tried it out in the `WatchStream` PR, see [this commit](https://github.com/firebase/firebase-ios-sdk/pull/1770/commits/a6d5da1b3ef14bf016eeff73977d5a204ff42ef7) for a diff.